# Tarea 1: Explorando datos mundiales de COVID-19 

En esta actividad se pide que explore los datos del repositorio ["COVID-19 Data Repository by the Center for Systems Science and Engineering (CSSE) at Johns Hopkins University"](https://github.com/CSSEGISandData/COVID-19)

Adjunto a este cuadernillo encontrará los datos al 26 de Abril de 2021 de dicho repositorio en el archivo `data.zip`. Luego de descomprimir debería observar cuatro archivos

- `covid19_confirmados.csv` : Número de casos diarios acumulado reportados como COVID-19 positivos (confirmados)
- `covid19_recuperados.csv` : Número de casos diarios acumulado reportados como recuperados 
- `covid19_decesos.csv` : Número de decesos diarios por COVID-19 acumulado 
- `population_by_country_2020.csv`:  Población por país más otros indicadores socioeconómicos al 2020 descargados desde: https://www.kaggle.com/tanuprabhu/population-by-country-2020

En los primeros tres archivos cada fila corresponde a un país o un estado/provincia de un país. Las columnas son las coordenadas geográficas (latitud y longitud) y el número de casos diarios para cada día desde Enero de 2020 a la fecha. Según quienes compilaron este repositorio, la información de casos recuperados debe analizarse con cautela pues podría estar considerablemente subestimada. 

A modo de ejemplo se muestra a continuación una gráfica de la series de tiempo del número de casos confirmados totales acumulados de Chile, Argentina y Bolivia 

![](https://i.imgur.com/7YVYk11.png)

A partir de esta información se puede calcular la serie de tiempo de los casos confirmados "nuevos" por día

![](https://i.imgur.com/baWnqxt.png)

En esta tarea se pide que procese, visualice, analice y discuta en base a estos datos. Escriba las rutinas de Python necesarias para resolver las actividades específicas que encontrará en este enunciado. **Use las librerías vistas en el curso**. Discuta con sus compañeros de grupo y complete con su análisis donde corresponda.

## Instrucciones generales 

1. Forme un grupo de **máximo tres estudiantes**
1. Versione su trabajo usando un **repositorio privado de github**. Agregue a sus compañeros y a su profesor (usuario github: phuijse) en la pestaña *Settings/Manage access*. No se aceptarán consultas de programación si no se cumple este requisito
1. Su tarea se evaluará en base al último commit antes de la fecha de entrega: **14:10 del Martes 11 de Mayo de 2021**. La nota se calcula como ("pt totales" + 1)
1. [Sean leales y honestos](https://www.acm.org/about-acm/code-of-ethics-in-spanish), no copie ni comparta resultados con otros grupos

## Enunciado

### Preparación de datos (1.5pt)

1. Explore el archivo `covid19_confirmados.csv` y configure la función `read_csv` de pandas para importarlo como un `DataFrame`. 
    1. Utilice el nombre de país como índice principal
    1. Elimine las columnas `Lat` y `Long`
    1. Use TimeStamps para los "nombres de columna" asociados a fechas
    1. Reduzca la columna `Province/State`, el DataFrame resultante debe tener los totales a nivel país
1. Explore el archivo `population_by_country_2020.csv` y configure la función `read_csv` de pandas para importarlo como un `DataFrame`. 
    1. Utilice el nombre de país como índice principal. Algunos siglas o nombres podrían no calzar con el `DataFrame` anterior, modifíquelos manualmente o usando pandas. 
1. Haga un `merge` de los `DataFrame` anteriores. El objetivo es asignar un valor de `Population` a cada elemento del primer `DataFrame`. Indique las filas donde no se puede realizar el `merge` y luego descártelas de su `DataFrame` final
1. Repita el procedimiento anterior para los archivos `covid19_recuperados.csv` y `covid19_muertes.csv` 

----

In [2]:
import matplotlib.pyplot as plt
import numpy as np 
import pandas as pd
import ipywidgets as widgets
import datetime

#### 1. Explore el archivo `covid19_confirmados.csv` y configure la función `read_csv` de pandas para importarlo como un `DataFrame`. 
    A. Utilice el nombre de país como índice principal
    B. Elimine las columnas `Lat` y `Long`
    C. Use TimeStamps para los "nombres de columna" asociados a fechas
    D. Reduzca la columna `Province/State`, el DataFrame resultante debe tener los totales a nivel país


In [3]:
def formatData(path):
    #A. Utilizamos el nombre de país como índice principal
    df = pd.read_csv(path, index_col ='Country/Region')
    df.index.name = 'Pais'
    #B. Eliminamos las columnas Lat y Long
    df.drop(['Lat', 'Long'], axis = 1 ,inplace = True)
    #C. Usamos TimeStamps para los "nombres de columna" asociados a fechas
    lista_col = [df.columns[0]]
    #D. Reducimos la columna Province/State
    dict_dias_sum = {}                                            
    for i in range(1,df.shape[1],1):
        lista_col.append(pd.Timestamp(df.columns[i]).date()) #C. 
        dict_dias_sum[lista_col[i]] = np.sum # D.
    df.columns = lista_col #C.
    df = df.groupby('Pais').aggregate(dict_dias_sum) #D.
    return df

In [4]:
confirmados = formatData('data/covid19_confirmados.csv')
display(confirmados)

2020-01-22  2020-01-23  2020-01-24  2020-01-25  \
Pais                                                                 
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      0           2           2           2   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-26  2020-01-27  2020-01-28  2020-01-29  \
Pais                                                                 
Afghanistan                  0           0           0           0   
Albania                      0           0           0           0   
Algeria                      0           0           0           0   
Andorra                      0           0           0           0   
Angola                       0           0           0           0   
...                        ...         ...         ...         ...   
Vietnam                      2           2           2           2   
West Bank and Gaza           0           0           0           0   
Yemen                        0           0           0           0   
Zambia                       0           0           0           0   
Zimbabwe                     0           0           0           0   

                    2020-01-30  2020-01-31  ...  2021-04-16  2021-04-17  \
Pais                                        ...                           
Afghanistan                  0           0  ...       57612       57721   
Albania                      0           0  ...      129307      129456   
Algeria                      0           0  ...      119323      119486   
Andorra                      0           0  ...       12712       12771   
Angola                       0           0  ...       24122       24300   
...                        ...         ...  ...         ...         ...   
Vietnam                      2           2  ...        2772        2781   
West Bank and Gaza           0           0  ...      278135      279753   
Yemen                        0           0  ...        5715        5770   
Zambia                       0           0  ...       90750       90844   
Zimbabwe                     0           0  ...       37534       37699   

                    2021-04-18  2021-04-19  2021-04-20  2021-04-21  \
Pais                                                                 
Afghanistan              57793       57898       58037       58214   
Albania                 129594      129694      129842      129980   
Algeria                 119642      119805      119992      120174   
Andorra                  12805       12805       12874       12917   
Angola                   24389       24518       24661       24883   
...                        ...         ...         ...         ...   
Vietnam                   2785        2791        2800        2812   
West Bank and Gaza      280741      282270      284280      286028   
Yemen                     5812        5858        5918        5960   
Zambia                   90918       90942       91042       91119   
Zimbabwe                 37751       37859       37875       37980   

                    2021-04-22  2021-04-23  2021-04-24  2021-04-25  
Pais                                                                
Afghanistan              58312       58542       58730       58843  
Albania                 130114      130270      130409      130537  
Algeria                 120363     

#### 2. Explore el archivo `population_by_country_2020.csv` y configure la función `read_csv` de pandas para importarlo como un `DataFrame`. 
    A. Utilice el nombre de país como índice principal. Algunos siglas o nombres podrían no calzar con el `DataFrame` anterior, modifíquelos manualmente o usando pandas. 


In [5]:
population = pd.read_csv('data/population_by_country_2020.csv', index_col = 'Country (or dependency)')
population.index.name = 'Pais'
display(population.columns)
population.drop(['Yearly Change', 'Net Change','Density (P/Km²)','Land Area (Km²)',
                'Migrants (net)', 'Fert. Rate', 'Med. Age','Urban Pop %', 'World Share'], 
                 axis = 1 ,inplace = True)
display("Población: ", population)

Index(['Population (2020)', 'Yearly Change', 'Net Change', 'Density (P/Km²)',
       'Land Area (Km²)', 'Migrants (net)', 'Fert. Rate', 'Med. Age',
       'Urban Pop %', 'World Share'],
      dtype='object')

'Población: '

Population (2020)
Pais                               
China                    1440297825
India                    1382345085
US                        331341050
Indonesia                 274021604
Pakistan                  221612785
...                             ...
Montserrat                     4993
Falkland Islands               3497
Niue                           1628
Tokelau                        1360
Holy See                        801

[235 rows x 1 columns]

#### 3.  Haga un `merge` de los `DataFrame` anteriores. El objetivo es asignar un valor de `Population` a cada elemento del primer `DataFrame`. Indique las filas donde no se puede realizar el `merge` y luego descártelas de su `DataFrame` final

In [6]:
def mg(df, population):
    return pd.merge(df, population, how = 'inner', on= 'Pais')

In [7]:
confirmados = mg(confirmados, population)
display("Casos Confirmados: " ,confirmados)

'Casos Confirmados: '

2020-01-22  2020-01-23  2020-01-24  2020-01-25  2020-01-26  \
Pais                                                                      
Afghanistan           0           0           0           0           0   
Albania               0           0           0           0           0   
Algeria               0           0           0           0           0   
Andorra               0           0           0           0           0   
Angola                0           0           0           0           0   
...                 ...         ...         ...         ...         ...   
Venezuela             0           0           0           0           0   
Vietnam               0           2           2           2           2   
Yemen                 0           0           0           0           0   
Zambia                0           0           0           0           0   
Zimbabwe              0           0           0           0           0   

             2020-01-27  2020-01-28  2020-01-29  2020-01-30  2020-01-31  ...  \
Pais                                                                     ...   
Afghanistan           0           0           0           0           0  ...   
Albania               0           0           0           0           0  ...   
Algeria               0           0           0           0           0  ...   
Andorra               0           0           0           0           0  ...   
Angola                0           0           0           0           0  ...   
...                 ...         ...         ...         ...         ...  ...   
Venezuela             0           0           0           0           0  ...   
Vietnam               2           2           2           2           2  ...   
Yemen                 0           0           0           0           0  ...   
Zambia                0           0           0           0           0  ...   
Zimbabwe              0           0           0           0           0  ...   

             2021-04-17  2021-04-18  2021-04-19  2021-04-20  2021-04-21  \
Pais                                                                      
Afghanistan       57721       57793       57898       58037       58214   
Albania          129456      129594      129694      129842      129980   
Algeria          119486      119642      119805      119992      120174   
Andorra           12771       12805       12805       12874       12917   
Angola            24300       24389       24518       24661       24883   
...                 ...         ...         ...         ...         ...   
Venezuela        181903      183190      184595      185736      186745   
Vietnam            2781        2785        2791        2800        2812   
Yemen              5770        5812        5858        5918        5960   
Zambia            90844       90918       90942       91042       91119   
Zimbabwe          37699       37751       37859       37875       37980   

             2021-04-22  2021-04-23  2021-04-24  2021-04-25  Population (2020)  
Pais                                                                            
Afghanistan       58312       58542       58730       58843           39074280  
Albania          130114      130270      130409      130537            2877239  
Algeria          120363      120562      120736      120922           43984569  
Andorra           12942       13007       13024       13060              77287  
Angola            25051       25279       25492       25609           33032075  
...                 ...         ...         ...         ...                ...  
Venezuela        188063      189381      190523      191518           28421581  
Vietnam            2824        2830        2833        2843           97490013  
Yemen              6020        6056        6105        6137           29935468  
Zambia            91189       91251       91317       91358           18468257  
Zimbabwe          38018       38045       38064 

#### 4.Repita el procedimiento anterior para los archivos `covid19_recuperados.csv` y `covid19_muertes.csv`

In [8]:
recuperados = mg(formatData('data/covid19_recuperados.csv'),population)
decesos = mg(formatData('data/covid19_decesos.csv'),population)
display("Recuperados: ", recuperados)
display("Decesos: ",decesos)

'Recuperados: '

2020-01-22  2020-01-23  2020-01-24  2020-01-25  2020-01-26  \
Pais                                                                      
Afghanistan           0           0           0           0           0   
Albania               0           0           0           0           0   
Algeria               0           0           0           0           0   
Andorra               0           0           0           0           0   
Angola                0           0           0           0           0   
...                 ...         ...         ...         ...         ...   
Venezuela             0           0           0           0           0   
Vietnam               0           0           0           0           0   
Yemen                 0           0           0           0           0   
Zambia                0           0           0           0           0   
Zimbabwe              0           0           0           0           0   

             2020-01-27  2020-01-28  2020-01-29  2020-01-30  2020-01-31  ...  \
Pais                                                                     ...   
Afghanistan           0           0           0           0           0  ...   
Albania               0           0           0           0           0  ...   
Algeria               0           0           0           0           0  ...   
Andorra               0           0           0           0           0  ...   
Angola                0           0           0           0           0  ...   
...                 ...         ...         ...         ...         ...  ...   
Venezuela             0           0           0           0           0  ...   
Vietnam               0           0           0           0           0  ...   
Yemen                 0           0           0           0           0  ...   
Zambia                0           0           0           0           0  ...   
Zimbabwe              0           0           0           0           0  ...   

             2021-04-17  2021-04-18  2021-04-19  2021-04-20  2021-04-21  \
Pais                                                                      
Afghanistan       52116       52168       52244       52272       52301   
Albania          101142      101584      102171      102601      103066   
Algeria           83286       83397       83514       83636       83765   
Andorra           12159       12203       12203       12285       12334   
Angola            22576       22597       22600       22647       22882   
...                 ...         ...         ...         ...         ...   
Venezuela        164530      165959      166826      168418      169685   
Vietnam            2475        2475        2475        2490        2490   
Yemen              2209        2219        2261        2281        2309   
Zambia            88555       88718       88860       88900       88956   
Zimbabwe          35005       35019       35031       35058       35065   

             2021-04-22  2021-04-23  2021-04-24  2021-04-25  Population (2020)  
Pais                                                                            
Afghanistan       52348       52363       52392       52489           39074280  
Albania          103582      104278      105016      105728            2877239  
Algeria           83900       84038       84167       84299           43984569  
Andorra           12375       12423       12458       12491              77287  
Angola            22901       23089       23092       23092           33032075  
...                 ...         ...         ...         ...                ...  
Venezuela        171043      172035      173097      173097           28421581  
Vietnam            2490        2490        2490        2516           97490013  
Yemen              2393        2454        2486        2546           29935468  
Zambia            89117       89256       89403       89574           18468257  
Zimbabwe          35073       35094       35101 

'Decesos: '

2020-01-22  2020-01-23  2020-01-24  2020-01-25  2020-01-26  \
Pais                                                                      
Afghanistan           0           0           0           0           0   
Albania               0           0           0           0           0   
Algeria               0           0           0           0           0   
Andorra               0           0           0           0           0   
Angola                0           0           0           0           0   
...                 ...         ...         ...         ...         ...   
Venezuela             0           0           0           0           0   
Vietnam               0           0           0           0           0   
Yemen                 0           0           0           0           0   
Zambia                0           0           0           0           0   
Zimbabwe              0           0           0           0           0   

             2020-01-27  2020-01-28  2020-01-29  2020-01-30  2020-01-31  ...  \
Pais                                                                     ...   
Afghanistan           0           0           0           0           0  ...   
Albania               0           0           0           0           0  ...   
Algeria               0           0           0           0           0  ...   
Andorra               0           0           0           0           0  ...   
Angola                0           0           0           0           0  ...   
...                 ...         ...         ...         ...         ...  ...   
Venezuela             0           0           0           0           0  ...   
Vietnam               0           0           0           0           0  ...   
Yemen                 0           0           0           0           0  ...   
Zambia                0           0           0           0           0  ...   
Zimbabwe              0           0           0           0           0  ...   

             2021-04-17  2021-04-18  2021-04-19  2021-04-20  2021-04-21  \
Pais                                                                      
Afghanistan        2539        2539        2546        2549        2557   
Albania            2340        2342        2347        2353        2358   
Algeria            3152        3155        3160        3165        3172   
Andorra             123         123         123         123         123   
Angola              561         561         563         565         570   
...                 ...         ...         ...         ...         ...   
Venezuela          1888        1905        1925        1944        1965   
Vietnam              35          35          35          35          35   
Yemen              1119        1126        1132        1138        1147   
Zambia             1234        1235        1236        1236        1238   
Zimbabwe           1552        1553        1553        1554        1555   

             2021-04-22  2021-04-23  2021-04-24  2021-04-25  Population (2020)  
Pais                                                                            
Afghanistan        2561        2565        2572        2582           39074280  
Albania            2364        2367        2372        2378            2877239  
Algeria            3181        3190        3198        3207           43984569  
Andorra             123         123         124         124              77287  
Angola              572         574         577         579           33032075  
...                 ...         ...         ...         ...                ...  
Venezuela          1987        2009        2028        2047           28421581  
Vietnam              35          35          35          35           97490013  
Yemen              1157        1165        1175        1187           29935468  
Zambia             1240        1245        1245        1246           18468257  
Zimbabwe           1555        1556        1556 

---

### Análisis global (1.5pt)

1. Obtenga la cantidad total  a la fecha (26 de Abril) de los casos (a) confirmados, (b) cerrados (recuperados más decesos) y (c) activos (confirmados menos cerrados) por país. En base a estas cantidades calcule el porcentaje de 
    1. casos activos con respecto al total de confirmados
    1. decesos con respecto al total de casos cerrados
    1. recuperados con respecto al total de casos cerrados
1. Luego calcule las siguientes tasas en base a información a la fecha
    1. Tasa de incidencia: confirmados por 100.000 dividido población total
    1. Tasa de recuperación: recuperados por 100.000 dividido población total
    1. Tasa de mortalidad: decesos por 100.000 dividido población total
1. En base a lo anterior muestre los nombres y los valores de los 5 países con
    1. La mayor y menor tasa de incidencia, respectivamente
    1. La mayor y menor tasa de recuperación, respectivamente
    1. La mayor y menor tasa de mortalidad, respectivamente
1. Responda, analice y discuta: ¿En qué lugar/ranking se encuentra Chile con respecto a cada una de las métricas de los puntos anteriores? ¿A qué países de la [OCDE](https://es.wikipedia.org/wiki/Organizaci%C3%B3n_para_la_Cooperaci%C3%B3n_y_el_Desarrollo_Econ%C3%B3micos) se parece más? ¿A qué paises de latinoamérica se parece más? 

#### 1. Obtenga la cantidad total  a la fecha (26 de Abril) de los casos (a) confirmados, (b) cerrados (recuperados más decesos) y (c) activos (confirmados menos cerrados) por país. En base a estas cantidades calcule el porcentaje de 
    A. casos activos con respecto al total de confirmados
    B. decesos con respecto al total de casos cerrados
    C. recuperados con respecto al total de casos cerrados

In [9]:
#(a)Confirmados
confirmados_totales = confirmados.iloc[:, 0:460].sum(axis =1)
display('Confirmados Totales' ,confirmados_totales)
#(b)Cerrados:recuperados más decesos
cerrados = recuperados.iloc[:, 0:460].sum(axis =1)+decesos.iloc[:, 0:460].sum(axis =1)
display('Recuperados Totales' ,cerrados)
#(c)Activos: confirmados menos cerrados
activos_totales = confirmados_totales - cerrados
display('Activos Totales', activos_totales)

'Confirmados Totales'

Pais
Afghanistan    14880160
Albania        15290074
Algeria        23443583
Andorra         1922686
Angola          3754670
                 ...   
Venezuela      28763642
Vietnam          476360
Yemen            741085
Zambia         10350726
Zimbabwe        5175628
Length: 180, dtype: int64

'Recuperados Totales'

Pais
Afghanistan    12180605
Albania        10163097
Algeria        16725608
Andorra         1747046
Angola          2912780
                 ...   
Venezuela      26412931
Vietnam          405154
Yemen            597904
Zambia          9708423
Zimbabwe        4550376
Length: 180, dtype: int64

'Activos Totales'

Pais
Afghanistan    2699555
Albania        5126977
Algeria        6717975
Andorra         175640
Angola          841890
                ...   
Venezuela      2350711
Vietnam          71206
Yemen           143181
Zambia          642303
Zimbabwe        625252
Length: 180, dtype: int64

    A. casos activos con respecto al total de confirmados

In [10]:
display('Porcentaje Activos respecto a Confirmados Totales', (activos_totales/confirmados_totales)*100)

'Porcentaje Activos respecto a Confirmados Totales'

Pais
Afghanistan    18.141976
Albania        33.531407
Algeria        28.655923
Andorra         9.135137
Angola         22.422477
                 ...    
Venezuela       8.172508
Vietnam        14.947939
Yemen          19.320456
Zambia          6.205391
Zimbabwe       12.080698
Length: 180, dtype: float64

    B. decesos con respecto al total de casos cerrados

In [11]:
decesos_totales = decesos.iloc[:, 0:460].sum(axis =1)
display('Porcentaje decesos respecto a cerrados totales', (decesos_totales/cerrados)*100)

'Porcentaje decesos respecto a cerrados totales'

Pais
Afghanistan     4.733000
Albania         2.889454
Algeria         4.308824
Andorra         1.591029
Angola          3.230213
                 ...    
Venezuela       1.011440
Vietnam         2.204841
Yemen          32.498361
Zambia          1.680427
Zimbabwe        4.033842
Length: 180, dtype: float64

    C. recuperados con respecto al total de casos cerrados

In [12]:
recuperados_totales = recuperados.iloc[:, 0:460].sum(axis =1)
display('Porcentaje Recuperados con respecto a Casos Cerrados',(recuperados_totales/cerrados)*100)

'Porcentaje Recuperados con respecto a Casos Cerrados'

Pais
Afghanistan    95.267000
Albania        97.110546
Algeria        95.691176
Andorra        98.408971
Angola         96.769787
                 ...    
Venezuela      98.988560
Vietnam        97.795159
Yemen          67.501639
Zambia         98.319573
Zimbabwe       95.966158
Length: 180, dtype: float64

---

#### 2. Luego calcule las siguientes tasas en base a información a la fecha
    A. Tasa de incidencia: confirmados por 100.000 dividido población total
    B. Tasa de recuperación: recuperados por 100.000 dividido población total
    C. Tasa de mortalidad: decesos por 100.000 dividido población total

In [13]:
poblacion_total = confirmados['Population (2020)']

In [14]:
#A.
tasa_incidencia = confirmados_totales*100000/poblacion_total
display("Tasa de Incidencia", tasa_incidencia)
#B.
tasa_recuperacion = recuperados_totales*100000/poblacion_total
display("Tasa de Recuperación", tasa_recuperacion)
#C.
tasa_mortalidad = decesos_totales*100000/poblacion_total
display("Tasa de Mortalidad", tasa_mortalidad)

'Tasa de Incidencia'

Pais
Afghanistan    3.808173e+04
Albania        5.314148e+05
Algeria        5.329956e+04
Andorra        2.487722e+06
Angola         1.136674e+04
                   ...     
Venezuela      1.012035e+05
Vietnam        4.886244e+02
Yemen          2.475609e+03
Zambia         5.604604e+04
Zimbabwe       3.473629e+04
Length: 180, dtype: float64

'Tasa de Recuperación'

Pais
Afghanistan    2.969753e+04
Albania        3.430177e+05
Algeria        3.638760e+04
Andorra        2.224501e+06
Angola         8.533194e+03
                   ...     
Venezuela      9.199270e+04
Vietnam        4.064221e+02
Yemen          1.348217e+03
Zambia         5.168479e+04
Zimbabwe       2.930797e+04
Length: 180, dtype: float64

'Tasa de Mortalidad'

Pais
Afghanistan     1475.415542
Albania        10206.242860
Algeria         1638.476894
Andorra        35964.651235
Angola           284.841325
                   ...     
Venezuela        939.958266
Vietnam            9.162990
Yemen            649.092909
Zambia           883.369773
Zimbabwe        1231.931685
Length: 180, dtype: float64

#### 3. En base a lo anterior muestre los nombres y los valores de los 5 países con
    A. La mayor y menor tasa de incidencia, respectivamente:

In [15]:
display("Mayor Tasa de incidencia",tasa_incidencia.sort_values(ascending=False)[:5])
display("Menor Tasa de incidencia",tasa_incidencia.sort_values()[:5])

'Mayor Tasa de incidencia'

Pais
Andorra       2.487722e+06
San Marino    2.001187e+06
Montenegro    1.853711e+06
Czechia       1.697898e+06
Bahrain       1.589746e+06
dtype: float64

'Menor Tasa de incidencia'

Pais
Micronesia     82.443093
Vanuatu        87.566526
Laos          170.822352
Samoa         190.794541
Tanzania      312.893669
dtype: float64

    B. La mayor y menor tasa de recuperación, respectivamente:
    

In [16]:
display("Mayor Tasa de Recuperación:",tasa_recuperacion.sort_values(ascending=False)[:5])
display("Menor Tasa de Recuperación:",tasa_recuperacion.sort_values()[:5])

'Mayor Tasa de Recuperación:'

Pais
Andorra       2.224501e+06
San Marino    1.686908e+06
Montenegro    1.557123e+06
Bahrain       1.485918e+06
Qatar         1.460042e+06
dtype: float64

'Menor Tasa de Recuperación:'

Pais
Sweden          0.000000
Vanuatu        48.648070
Micronesia     75.500516
Tanzania      110.818764
Laos          140.411888
dtype: float64

    C. La mayor y menor tasa de mortalidad, respectivamente.

In [17]:
display("Mayor Tasa de Mortalidad:",tasa_mortalidad.sort_values(ascending=False)[:5])
display("Menor Tasa de Mortalidad:",tasa_mortalidad.sort_values()[:5])

'Mayor Tasa de Mortalidad:'

Pais
San Marino        60682.300259
Belgium           47047.541090
United Kingdom    36175.289907
Italy             36079.474857
Andorra           35964.651235
dtype: float64

'Menor Tasa de Mortalidad:'

Pais
Laos                0.0
Samoa               0.0
Solomon Islands     0.0
Dominica            0.0
Marshall Islands    0.0
dtype: float64

----

### 4. Responda, analice y discuta: 

    ¿En qué lugar/ranking se encuentra Chile con respecto a cada una de las métricas de los puntos anteriores? 

In [18]:
idx_chile = tasa_incidencia.sort_values(ascending=False).index.get_loc("Chile")
display("Ranking Tasa de incidencia:",tasa_incidencia.sort_values(ascending=False)[:idx_chile+2])
print("Lugar: ", idx_chile)

'Ranking Tasa de incidencia:'

Pais
Andorra          2.487722e+06
San Marino       2.001187e+06
Montenegro       1.853711e+06
Czechia          1.697898e+06
Bahrain          1.589746e+06
Qatar            1.578398e+06
Luxembourg       1.565217e+06
US               1.488810e+06
Israel           1.437300e+06
Panama           1.407219e+06
Slovenia         1.305746e+06
Belgium          1.243815e+06
Armenia          1.183300e+06
Spain            1.127049e+06
Portugal         1.112384e+06
Switzerland      1.072490e+06
Sweden           1.063131e+06
Netherlands      1.053932e+06
Liechtenstein    1.047765e+06
Kuwait           1.017121e+06
Lithuania        1.012388e+06
France           1.009385e+06
Georgia          1.006302e+06
Brazil           1.000630e+06
Chile            9.930805e+05
Croatia          9.463632e+05
dtype: float64

Lugar:  24


In [19]:
idx_chile = tasa_mortalidad.sort_values(ascending=False).index.get_loc("Chile")
display("Ranking Tasa de Mortalidad:", tasa_mortalidad.sort_values(ascending=False)[:idx_chile+2])
print("Lugar: ", idx_chile)

'Ranking Tasa de Mortalidad:'

Pais
San Marino                60682.300259
Belgium                   47047.541090
United Kingdom            36175.289907
Italy                     36079.474857
Andorra                   35964.651235
Spain                     35027.105534
Peru                      33251.013133
US                        31308.107462
France                    29224.166460
Sweden                    28215.581184
Czechia                   27811.017821
North Macedonia           27529.196840
Mexico                    27517.051929
Brazil                    27279.937609
Bosnia and Herzegovina    27135.055072
Slovenia                  26221.543260
Montenegro                26086.326583
Panama                    25301.713059
Chile                     24983.398717
Hungary                   24076.595827
dtype: float64

Lugar:  18


In [20]:
idx_chile = tasa_recuperacion.sort_values(ascending=False).index.get_loc("Chile")
display("Ranking Tasa de Recuperación", tasa_recuperacion.sort_values(ascending=False)[:idx_chile+2])
print("Lugar: ", idx_chile)

'Ranking Tasa de Recuperación'

Pais
Andorra          2.224501e+06
San Marino       1.686908e+06
Montenegro       1.557123e+06
Bahrain          1.485918e+06
Qatar            1.460042e+06
Czechia          1.438139e+06
Luxembourg       1.368835e+06
Israel           1.310646e+06
Panama           1.215443e+06
Slovenia         1.121572e+06
Armenia          1.015878e+06
Liechtenstein    9.623771e+05
Kuwait           9.356464e+05
Georgia          9.353575e+05
Chile            9.215240e+05
Portugal         9.161790e+05
dtype: float64

Lugar:  14


----

     ¿A qué países de la OCDE se parece más?

**Nota:** se cambiaron los nombres de: 
* United States → USA
* South Korea → Korea, South
* Czech Republic (Czechia) → Czechia
                                   
en `population_by_country_2020.csv`          

In [21]:
OCDE = ["Austria", "Australia", "Belgium", "Canada", "Chile", "Colombia", "Denmark", "Estonia", 
        "Finland", "France", "Germany", "Greece", "Hungary", "Iceland", "Ireland", "Israel", "Italy", 
        "Japan", "Latvia", "Lithuania", "Luxembourg", "Mexico", "Netherlands", "New Zealand", "Norway", 
        "Poland", "Portugal", "Slovenia", "Spain", "Sweden", 
        "Switzerland", "Turkey", "US","United Kingdom", "Slovakia", "Czechia"]
# Fuente: https://www.oecd.org/newsroom/global-oecd-welcomes-colombia-as-its-37th-member.htm

In [22]:
display("Tasa de Incidencia OCDE",tasa_incidencia[OCDE].sort_values(ascending=False))

'Tasa de Incidencia OCDE'

Pais
Czechia           1.697898e+06
Luxembourg        1.565217e+06
US                1.488810e+06
Israel            1.437300e+06
Slovenia          1.305746e+06
Belgium           1.243815e+06
Spain             1.127049e+06
Portugal          1.112384e+06
Switzerland       1.072490e+06
Sweden            1.063131e+06
Netherlands       1.053932e+06
Lithuania         1.012388e+06
France            1.009385e+06
Chile             9.930805e+05
United Kingdom    9.295139e+05
Austria           8.483485e+05
Italy             8.216103e+05
Colombia          8.070122e+05
Slovakia          7.814105e+05
Poland            7.513564e+05
Hungary           7.452721e+05
Estonia           7.206065e+05
Ireland           6.949607e+05
Latvia            5.846446e+05
Denmark           5.820028e+05
Turkey            5.298301e+05
Germany           4.813201e+05
Iceland           4.410277e+05
Canada            3.943661e+05
Greece            3.068237e+05
Mexico            2.940431e+05
Norway            2.444908e+05
Fin

In [23]:
display("Tasa de Mortalidad OCDE",tasa_mortalidad[OCDE].sort_values(ascending=False))

'Tasa de Mortalidad OCDE'

Pais
Belgium           47047.541090
United Kingdom    36175.289907
Italy             36079.474857
Spain             35027.105534
US                31308.107462
France            29224.166460
Sweden            28215.581184
Czechia           27811.017821
Mexico            27517.051929
Slovenia          26221.543260
Chile             24983.398717
Hungary           24076.595827
Portugal          22160.704177
Colombia          22114.103023
Switzerland       21050.992919
Netherlands       20820.283167
Luxembourg        18856.499242
Ireland           18462.516779
Poland            17327.722229
Slovakia          15977.190199
Lithuania         15378.000545
Austria           14693.155382
Canada            12834.412984
Germany           12473.851550
Israel            11032.276597
Latvia            10482.955874
Greece             9595.574622
Denmark            7422.072611
Estonia            7074.055565
Turkey             6342.070642
Finland            3147.952713
Norway             2623.036462
Ice

In [24]:
display("Tasa de Recuperación OCDE", tasa_recuperacion[OCDE].sort_values(ascending=False))

'Tasa de Recuperación OCDE'

Pais
Czechia           1.438139e+06
Luxembourg        1.368835e+06
Israel            1.310646e+06
Slovenia          1.121572e+06
Chile             9.215240e+05
Portugal          9.161790e+05
Lithuania         8.407734e+05
Austria           7.588820e+05
Colombia          7.300909e+05
Switzerland       6.599737e+05
Italy             6.020396e+05
Poland            5.876956e+05
Slovakia          5.693988e+05
Estonia           5.475944e+05
Denmark           5.187145e+05
Turkey            4.764537e+05
Latvia            4.722924e+05
Hungary           4.346293e+05
Iceland           4.107649e+05
Germany           4.096648e+05
Canada            3.445741e+05
Mexico            2.302707e+05
Ireland           1.688314e+05
US                1.499880e+05
Greece            1.458522e+05
Finland           1.260082e+05
Spain             1.197568e+05
France            8.691805e+04
Norway            8.477921e+04
Japan             4.642977e+04
Belgium           3.274411e+04
Australia         2.515535e+04
New

**COMENTARIOS:**

Para realizar este analisis, primero se selecciona con que paises existe una mayor simlitud en cuanto a su ranking, buscanso la cercania de su posición con Chile. Se consideran aquellos paises hasta 2 posiciones por encima y debajo de Chile (si aparece en uno pero en otro no, se considerará igualmente):

En la **Taza de incidencia**:
- Ukraine            3131.799280
- Indonesia          2732.994681
- Chile              2516.614001
- Czechia            2407.617508
- Netherlands        2391.377361

En la **Taza de mortalidad**:
- Turkey              70.933121
- Canada              64.216514
- Chile               63.311654
- Ukraine             60.438933
- Ecuador             51.347021

En la **Taza de recuperación**:
- Iran             3352.220731
- Poland           2943.613081
- Chile            2335.279133
- Indonesia        2284.383808
- Ukraine          2265.223247

Luego se ve los paises que más se repiten en cuanto a su cercania, tomando la distancia con Chile, buscando la que en promedio se encuentre mas cercana a Chile:

- **Ukraine**: 2 posiciones por encima en **incidencia**, 1 posición por debajo en **mortalidad** y 2 posiciones por debajo en **recuperación**.
    TOTAL: ((2+1+2)/3) = 1.66...
- **Indonesia**: 1 posición por encima en **incidencia**, 4 posiciones por encima en **mortalidad** y 1 posición por debajo en **recuperación**.
    TOTAL: ((1+4+1)/3) = 2
- **Czechia**: 1 posición por debajo en **incidencia**, 6 posciones por debajo en **mortalidad** y 3 posiciones por debajo en **recuperación**. TOTAL: ((1+6+3)/3) = 3.33..
- **Netherlands**: 2 posiciones por debajo de en **incidencia**, 4 posiciones por debajo en **mortalidad** y 102 posiciones por debajo en **recuperación**.
    TOTAL: ((2+4+102)/3) = 36
- **Turkey**: 11 posiciones por encima en **inicidencia**, 2 posiciones por encima en **mortalidad** y 10 posiciones por encima en **recuperación**.
    TOTAL: ((11+2+10)/3) = 7.66...
- **Canada**: 4 posiciones por debajo en **incidencia**, 1 posición por encima en **mortalidad** y 5 posiciones por debajo en **recuperación**.
    TOTAL: ((4+1+5)/3) = 3.33..
- **Ecuador**: 22 posiciones por debajo en **incidencia**, 2 posiciones por encima en **mortalidad** y 22 posiciones por debajo en **recuperación**.
    TOTAL: ((22+2+22)/3) = 15.33..
- **Iran**: 6 posiciones por encima en **incidencia**, 11 posiciones por encima en **mortalidad** y 2 posiciones por encima en **recuperación**.
    TOTAL: ((6+11+2)/3) = 6.33..
- **Poland**: 3 posiciones por encima en **incidencia**, 5 posiciones por encima en **mortalidad** y 1 posición por encima en **recuperación**.
    TOTAL ((3+5+1)/3) = 3

**Resultado:** Ukraine -> Indonesia -> Poland

Se calculará ahora la proporción por cuanto supera Chile al pais o el por cuanto el pais supera a Chile, con la formula:

- (((monto pais/monto chile)-1)*100)

1. Primer lugar: Ukraine con una **Taza de incidencia** aproximadamente **24.4% superior** a Chile, una **Taza de mortalidad** aproximadamente **4.7% inferior** a Chile y una **Taza de recuperación** aproximadamente **3% superior** a Chile.
2. Segundo lugar: Indonesia con una **Taza de incidencia** aproximadamente aproximadamente **8.5% superior** a Chile, una **Taza de mortalidad** aproximadamente un **28.6% superior** a Chile y una **Taza de recuperación** aprocimadamente **2.2% inferior** a Chile.
3. Tercer lugar: Poland con una **Taza de incidencia** aproximadamente **49.5% superior** a Chile, una **Taza de mortalidad** aproximadamente **37.1% superior** a Chile y una **Taza de recuperación** aproximadamente **26.1% superior** a Chile.

**Conclusión:** Chile se asemeja mayormente a Ukrania en cuanto a su comportamiento sobre estas 3 variables y sus números cencanos entre si.


Además, individualmente los valores mas cercanos a los de Chile son:
- La **Taza de incidencia** de Chile es un aproximado de **4.5% superior** a Czechia.
- La **Taza de mortalidad** de Chile es un aproximado de **1.4% inferior** a Canada.
- La **Taza de recuperación** de Chile es un aproximado **2.2% superior** a Indonesia.

-----

    ¿A qué paises de latinoamérica se parece más?

In [25]:
LATAM = ["Argentina", "Bolivia", "Brazil", "Chile", "Colombia", "Costa Rica", "Cuba", "Dominican Republic", 
         "Ecuador", "El Salvador", "Guatemala", "Honduras", "Mexico", "Nicaragua", "Panama", "Paraguay", 
         "Peru", "Haiti", "Uruguay", "Venezuela"]

# Fuente: https://en.wikipedia.org/wiki/Latin_Americans

In [26]:
display("Tasa de Incidencia LATAM", tasa_incidencia[LATAM].sort_values(ascending=False))

'Tasa de Incidencia LATAM'

Pais
Panama                1.407219e+06
Brazil                1.000630e+06
Chile                 9.930805e+05
Peru                  9.010308e+05
Argentina             8.700892e+05
Colombia              8.070122e+05
Costa Rica            7.438188e+05
Dominican Republic    4.384086e+05
Bolivia               4.291974e+05
Paraguay              3.885454e+05
Ecuador               3.453535e+05
Honduras              3.378897e+05
Mexico                2.940431e+05
Uruguay               2.643145e+05
Guatemala             2.040912e+05
El Salvador           1.880788e+05
Venezuela             1.012035e+05
Cuba                  6.188427e+04
Haiti                 2.693347e+04
Nicaragua             2.544510e+04
dtype: float64

In [27]:
display('Tasa de Mortalidad LATAM', tasa_mortalidad[LATAM].sort_values(ascending=False))

'Tasa de Mortalidad LATAM'

Pais
Peru                  33251.013133
Mexico                27517.051929
Brazil                27279.937609
Panama                25301.713059
Chile                 24983.398717
Colombia              22114.103023
Ecuador               21929.871326
Bolivia               21637.335134
Argentina             21553.989049
Costa Rica             9624.966534
Honduras               8888.343589
Paraguay               7928.209836
Guatemala              7243.193651
Dominican Republic     6663.459918
El Salvador            5540.530903
Uruguay                2893.601032
Venezuela               939.958266
Nicaragua               725.571796
Haiti                   604.532189
Cuba                    559.609818
dtype: float64

In [28]:
display('Tasa de Recuperacion LATAM', tasa_recuperacion[LATAM].sort_values(ascending=False))

'Tasa de Recuperacion LATAM'

Pais
Panama                1.215443e+06
Chile                 9.215240e+05
Brazil                8.692808e+05
Peru                  7.931803e+05
Argentina             7.520707e+05
Colombia              7.300909e+05
Costa Rica            5.592121e+05
Dominican Republic    3.360329e+05
Bolivia               3.140059e+05
Paraguay              3.020510e+05
Ecuador               2.809910e+05
Mexico                2.302707e+05
Uruguay               2.136519e+05
Guatemala             1.796724e+05
El Salvador           1.612242e+05
Honduras              1.275209e+05
Venezuela             9.199270e+04
Cuba                  5.573046e+04
Haiti                 2.025656e+04
Nicaragua             1.719267e+04
dtype: float64

**COMENTARIOS:**
     
Se usará el mismo criterio que con OCDE:

En la **Taza de incidencia**:
- Mexico                 5027.417791
- Peru                   3941.845404
- Chile                  2516.614001
- Ecuador                 808.617374
- Panama                  805.867412

En la **Taza de mortalidad**:
- Peru                  145.467124
- Argentina             129.151487
- Chile                  63.311654
- Ecuador                51.347021
- Bolivia                33.510665

En la **Taza de recuperación**:
- Mexico                 3937.066611
- Peru                   3470.019076
- Chile                  2335.279133
- Panama                  696.044011
- Ecuador                 657.917631

- **Mexico**:  2 posiciones por encima en **incidencia**, 4 posiciones por encima en **mortalidad** y 2 posiciones por encima en **recuperación**.
    TOTAL: ((2+4+2)/3) = 2.66..
- **Peru**:  1 posición por encima en **incidencia**, 2 posiciones por encima en **mortalidad** y 1 posición por encima en **recuperación**.
    TOTAL: ((1+2+1)/3) = 1.33..
- **Ecuador**: 1 posición por debajo en **incidencia**, 1 posición por debajo en **mortalidad** y 2 posiciones por debajo en **recuperación**.
    TOTAL: ((1+1+2)/3) = 1.33..
- **Panama**: 2 posiciones por debajo en **incidencia**, 4 posiciones por debajo en **mortalidad** y 1 posición por debajo en **recuperación**.
    TOTAL: ((2+4+1)/3) = 2.33..
- **Argentina**: 3 posiciones por encima en **incidencia**, 1 posición por encima en **mortalidad** y 3 posiciones por encima en **recuperación**.
    TOTAL: ((3+1+3)/3) = 2.33..
- **Bolivia**: 3 posiciones por debajo en **incidencia**, 2 posiciones por debajo en **mortalidad** y 3 posiciones por debajo en **recuperación**.
    TOTAL: ((3+2+3)/3) = 2.66..
    
**Resultado**: Peru/Ecuador -> Panama/Argentina -> Mexico/Bolivia

Se concerva la formula anterior:
- (((monto pais/monto chile)-1)*100)

1. Primer lugar
- Peru con una **Taza de incidencia** aproximadamente **56.6% superior** a Chile, una **Taza de mortalidad** aproximadamente **129.8% superior** a Chile y una **Taza de recuperación** aproximadamente **48.6% superior** a Chile.
- Ecuador con una **Taza de incidencia** aproximadamente **67.9% inferior** a Chile, una **Taza de mortalidad** aproximadamente **19.1% inferior** a Chile y una **Taza de recuperación** aproximadamente **71.9% inferior** a Chile.

2. Segundo lugar
- Panama con una **Taza de incidencia** aproximadamente **68% inferior** a Chile, una **Taza de mortalidad** aproximadamente **77.77..% inferior** a Chile y una **Taza de recuperación** aproximadamente **70.19% inferior** a Chile.
- Argentina con una **Taza de incidencia** aproximadamente **107.2% superior** a Chile, una **Taza de mortalidad** aproximadamente **104.7% superior** a Chile y una **Taza de recuperación** aproximadamente **92.8% superior** a Chile.

3. Tercer Lugar
- Mexico con una **Taza de incidencia** aproximadamente **99.8% superior** a Chile, una **Taza de mortalidad** aproximadamente **646% superior** a Chile y una **Taza de recuperación** aproximadamente **68.6% superior** a Chile.
- Bolivia con una **Taza de incidencia** aproximadamente **73.6% inferior** a Chile, una **Taza de mortalidad** aproximadamente **47.6% inferior** a Chile y una **Taza de recuperación** aproximadamente **79.2% inferior** a Chile.

**Conclusión:** Chile se asemeja mayormente a Ecuador en cuanto a su comportamiento sobre estas 3 variables y sus números cencanos entre si.

Además, individualmente los valores mas cercanos a los de Chile son:
- La **Taza de incidencia** de Chile es un aproximado de **56.6% superior** a Peru.
- La **Taza de mortalidad** de Chile es un aproximado de **19.1% inferior** a Ecuador.
- La **Taza de recuperación** de Chile es un aproximado **68.6% superior** a Mexico.

----

### Análisis temporal y comparativo (2.0pt)

1. Construya una interfaz gráfica usando `ipywidgets` y `matplotlib` que permita realizar comparaciones entre las series de tiempo de COVID-19 de dos países seleccionados arbitrariamente. A modo de simplificación considere sólo los países de la [OCDE](https://es.wikipedia.org/wiki/Organizaci%C3%B3n_para_la_Cooperaci%C3%B3n_y_el_Desarrollo_Econ%C3%B3micos). Requerimientos:
    1. La interfaz debe incluir controles para escoger dos países a partir de una lista
    1. La interfaz debe incluir tres [tabs](https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20List.html#Tabs), el primero debe mostrar las series de tiempo de ambos países de los casos confirmados, el segundo la de recuperados y el tercero la de los decesos, respectivamente. Use una leyenda para indicar el nombre de los países.
    1. La interfaz debe tener un control que permita escoger entre casos acumulados y casos nuevos  
    1. La interfaz debe tener un control que permita escoger entre valores absolutos y valores relativos (tasas por 100.000 habitantes)
    1. La interfaz debe tener un control que permita escoger entre valores diarios y valores semanales (acumulados). Para esto último se recomienda usar `groupby` con una [frecuencia lunes a lunes](https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#anchored-offsets)
    
En base a su interfaz:   

1. Describa en detalle la situación de Chile con respecto a la evolución temporal de las series de tiempo. En particular destaque patrones y comportamientos que considere relevantes y compare lo que observa en las distintas series de tiempo. Considerando la evolución a nivel de semana, conteste ¿En qué fechas ocurre un aumento o disminución notoria de los casos? ¿En qué fechas hay máximos y mínimos?
1. Considerando métricas que sean relativas compare el caso de Chile contra cinco países de la OCDE seleccionados por ustedes. Destaque semejanzas y diferencias considerando las cantidades y los posibles desfases temporales existentes.


In [29]:
def acum_pais(data,pais):
    pais_fecha = data.loc[pais].index
    pais_pers = data.loc[pais].values
    return (pais_fecha[0: len(pais_fecha)-1],pais_pers[0: len(pais_pers)-1])

def casos_nuevos(data, pais):
    casos_nuevos = np.zeros(shape = data.shape[0])
    casos = data.drop(['Population (2020)'], axis=1)
    casos = (casos.loc[pais]).values
    casos_nuevos = np.zeros(shape = casos.shape[0])
    ayer,hoy = 0,0
    d = 0
    tiempo = []
    m = 1
    fecha = data.loc[pais].index
    for dia in casos:
        tiempo.append(dia)
        hoy = np.sum(casos[m-1])
        casos_nuevos[d] = hoy - ayer
        ayer = hoy
        d+=1
        m+=1
    return (fecha[0: len(fecha)-1],casos_nuevos)

def generate_plot(data,pais1,pais2,caso,titulo):
    fig, ax = plt.subplots(figsize=(12, 6), tight_layout=True)
    if(caso =="Acumulados"):
        (pais1_x, pais1_y) = acum_pais(data,pais1)
        (pais2_x, pais2_y) = acum_pais(data,pais2)               
    else:
        (pais1_x, pais1_y) = casos_nuevos(data,pais1)
        (pais2_x, pais2_y) = casos_nuevos(data,pais2)
                  
    ax.plot(pais1_x,pais1_y, label = pais1);ax.legend()
    ax.plot(pais2_x,pais2_y ,label = pais2);ax.legend()
    ax.set(xlabel="Fecha", ylabel="Personas")
    ax.set_title(titulo,fontsize=18)
    plt.show()
    
def generate_plot_confirmados(pais1,pais2,caso):
    generate_plot(confirmados,pais1,pais2,caso, "Confirmados")
def generate_plot_recuperados(pais1,pais2,caso):
    generate_plot(recuperados,pais1,pais2,caso, "Recuperados")
def generate_plot_decesos(pais1,pais2,caso):
    generate_plot(decesos,pais1,pais2,caso, "Decesos")
    

In [30]:
pais1_dropdown = widgets.Dropdown(options=list(OCDE),
                                description='País:',
                                disabled=False)

pais2_dropdown = widgets.Dropdown(options=list(OCDE),
                                description='País:',
                                disabled=False)

casos_slider = widgets.Dropdown(description="Casos:", 
                              options=["Nuevos", "Acumulados"])
def action_1():
    widgets.interact(generate_plot_confirmados, pais1=pais1_dropdown, pais2=pais2_dropdown, caso=casos_slider);

def action_2():
    widgets.interact(generate_plot_recuperados, pais1=pais1_dropdown, pais2=pais2_dropdown,caso =casos_slider);

def action_3():
    widgets.interact(generate_plot_decesos, pais1=pais1_dropdown, pais2=pais2_dropdown,caso =casos_slider);

out1 = widgets.interactive(action_1)
out2 = widgets.interactive(action_2)
out3 = widgets.interactive(action_3)

tab  = widgets.Tab(children = [out1, out2, out3])
tab.set_title(0, 'Confirmados')
tab.set_title(1, 'Recuperados')
tab.set_title(2, 'Decesos')

display(tab)


### Dashboard interactivo (1.0pt)

Implemente un *dashboard web* interactivo a partir de la interfaz desarrollada en el punto anterior utilizando la librería [voila](https://github.com/voila-dashboards/voila). Sirva su interfaz en la nube, para esto puede considerar los servicios gratuitos [binder](https://mybinder.org/) o [heroku](https://www.heroku.com/free). Revise la documentación de voila [con respecto a *deployment* en la nube](https://voila.readthedocs.io/en/stable/deploy.html) y este repositorio con [un ejemplo en base a heroku](https://github.com/voila-dashboards/voila-heroku). Entregue el link de su dashboard web

[LINK HEROKU APP](https://desolate-beyond-69189.herokuapp.com/)